### PR TITLE
Feature/request model #9

### LINKED ISSUE #9: BE: Request Model
The following are properties you may want to consider adding to this model but is not limited to:

\- user\_id (user who is requesting)
\- sitter\_id (pet sitter user id)
\- start (date requested to start)
\- end (date requested to end)
\- accepted (default False) 
\- declined (default False)
\- paid (default False)

### PR DESCRIPTION
**Pull request by MarianaES**
### What this PR does (required):
- Implement a model for booking requests

### Any information needed to test this feature (required):
- These are just instructions for a collection to be created later.

Closes #9

### CODE REVIEW

File: server/models/Request.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  end: { type: Date, required: true },
+  accepted: { type: Boolean, default: false },
+  declined: { type: Boolean, default: false },
+  paid: { type: Boolean, default: false },
+  serviceType: {
+    type: String,
+    enum: ["Boarding", "House Sitting", "Day Care", "Walking"],
+  },
+  rating: { type: Number },
+});
```
Comment (ezeilo-su) [inline server/models/Request.js:18]: You can add a `timestamps` option here to let Mongoose automatically manage `createdAt` and `updatedAt` properties on the documents. Check it out [here](https://mongoosejs.com/docs/guide.html#timestamps)

**Review by ezeilo-su — COMMENTED** [review]
Looks good!

File: server/models/Request.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import mongoose from "mongoose";
+
+const { Schema, model } = mongoose;
+
+const requestSchema = new mongoose.Schema(
+  {
+    customer: { type: Schema.Types.ObjectId, ref: "User", required: true },
+    provider: { type: Schema.Types.ObjectId, ref: "User", required: true },
```
Comment (ugur-akin) [inline server/models/Request.js:8]: These are nice! I was wondering maybe `owner/sitter` could be clearer names for them? Any thoughts?

File: server/models/Request.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    start: { type: Date, required: true },
+    end: { type: Date, required: true },
+    accepted: { type: Boolean, default: false },
+    declined: { type: Boolean, default: false },
```
Comment (ugur-akin) [inline server/models/Request.js:12]: I think keeping two fields for tracking the status of a booking introduces unnecessary complexity and isn't the most future-proof solution. I'd recommend reducing this to a single field (maybe we can use `enum` for this too!).

File: server/models/Request.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    paid: { type: Boolean, default: false },
+    serviceType: {
+      type: String,
+      enum: ["Boarding", "House Sitting", "Day Care", "Walking"],
+    },
```
Comment (ugur-akin) [inline server/models/Request.js:17]: This is a nice addition to the basic schema, also nice use of the `enum` validator! 🙌🏼

One thing I would suggest would be to normalize values to either `lowercase/uppercase` to be more flexible in terms of casing ([docs](https://mongoosejs.com/docs/schematypes.html#string-validators)).

File: server/models/Request.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    totalPrice: { type: Number, required: true },
```
Comment (ugur-akin) [inline server/models/Request.js:18]: On one hand I think this is a nice addition. On the other hand, the most common practice for this would be, say, having an `Invoice` model and provide a `ref` here... But considering the scope of the project, it might be better to leave this in and not complicate things in favor of time/completion. Just wanted to share thoughts 👀 👍🏼 

Edit: I remembered there is a `paid` field here too so maybe it's best to do it this way for now 👍🏼

File: server/models/Request.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import mongoose from "mongoose";
+
+const { Schema, model } = mongoose;
+
+const requestSchema = new mongoose.Schema(
+  {
+    customer: { type: Schema.Types.ObjectId, ref: "User", required: true },
+    provider: { type: Schema.Types.ObjectId, ref: "User", required: true },
+    start: { type: Date, required: true },
+    end: { type: Date, required: true },
```
Comment (ugur-akin) [inline server/models/Request.js:10]: As a bonus, you could look into adding validation here to make sure `end > start` 😎

**Review by ugur-akin — CHANGES_REQUESTED** [review]
Awesome start to the project! 💯 I left a few thoughts/comments, and also if you are interested, you can take a look at adding more `SchemaType` options where appropriate as a bonus (the model is good enough as is, but I feel it might be a good exercise if we have the time, don't worry if not) 👍🏼 Let's check back once you can have a look and hopefully we'll be able to merge this soon!

One last thing about formatting PRs going forward, it would really helpful if you could:
- [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to the PR through github, 
- set self-explanatory PR titles in plain English (e.g. `Implement a model for booking requests`),
- add any relevant labels to the PR.

You can also have a look at this [article](https://hugooodias.medium.com/the-anatomy-of-a-perfect-pull-request-567382bb6067) that goes into more detail about how best to format PRs.

Thanks for all your work! 🙏🏼

File: server/models/Request.js (hunk 5 above)
Comment (MarianaES) [reply server/models/Request.js:18]: Thanks, the reason to include `totalPrice` arose from the payment feature where I think we'll need this amount. Another model for Invoice sounds better, but following your advice, I'm gonna leave it like that.

File: server/models/Request.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+const mongoose = require("mongoose");
+
+const { Schema, model } = mongoose;
+
+const requestSchema = new mongoose.Schema(
+  {
+    owner: { type: Schema.Types.ObjectId, ref: "User", required: true },
```
Comment (DestinTech) [inline server/models/Request.js:7]: Optional: Do you mind reformatting this to use multi-line instead of all on one line? It will allow refactoring to be easier in the future. 
```
  owner: {
    type: Schema.Types.ObjectId,
    ref: "User",
    required: true,
  },
```

File: server/models/Request.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    sitter: { type: Schema.Types.ObjectId, ref: "User", required: true },
+    startDate: { type: Date, required: true },
```
Comment (DestinTech) [inline server/models/Request.js:9]: Same for these if possible, please!

File: server/models/Request.js (hunk 7 above)
Comment (MarianaES) [reply server/models/Request.js:7]: Yes, no problem. Prettier accepted both ways 😅

**Review by DestinTech — COMMENTED** [review]
Looks good here! Great job making these changes.

File: server/models/Request.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      type: Number,
+    },
+    rating: {
+      type: Number,
+    },
+  },
+  { timestamps: true }
+);
+
+requestSchema.pre("validate", function (next) {
```
Comment (rajivtitus) [inline server/models/Request.js:43]: 👍🏽

File: server/models/Request.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      type: Date,
+      required: true,
+    },
+    endDate: {
+      type: Date,
+      required: true,
+    },
+    status: {
+      type: String,
+      enum: ["ACCEPTED", "DECLINED", "PAID"],
```
Comment (rajivtitus) [inline server/models/Request.js:27]:
```suggestion
      enum: ["ACCEPTED", "DECLINED", "PAID", "PENDING"],
      required: true,
      default: "PENDING"
```

All the fields in this model will be 'required', please update accordingly :)

File: server/models/Request.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    },
+    serviceType: {
+      type: String,
+      enum: ["BOARDING", "HOUSE_SITTING", "DAY_CARE", "WALKING"],
```
Comment (rajivtitus) [inline server/models/Request.js:31]:
```suggestion
      enum: ["BOARDING", "HOUSE_SITTING", "DAY_CARE", "WALKING"],
      uppercase: true,
```

**Review by rajivtitus — COMMENTED** [review]
Good work here! Some minor comments for you to address, thanks! :)

File: server/models/Request.js (hunk 10 above)
Comment (MarianaES) [reply server/models/Request.js:27]: I noticed that a pending status was missing by doing the request controller PR, so I added it there. 😅 However, I can include it here too as you suggest.

**Review by rajivtitus — APPROVED** [review]
Great work implementing those changes 💯
